assets/src/dashboard/components/storyAnimation/WAAPIWrapper.js

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/**
+ * External dependencies
+ */
+import { useMemo } from 'react';
+import PropTypes from 'prop-types';
+/**
+ * Internal dependencies
+ */
+import { WAAPI } from '../../animations/animationParts';
+import useStoryAnimationContext from './useStoryAnimationContext';
+
+function ComposableWrapper({ generators, hoistAnimation, children }) {
+  const ComposedWrapper = useMemo(
+    () =>
+      generators.reduce(
+        (Composable, generate) => {
+          // eslint-disable-next-line @wordpress/no-unused-vars-before-return
+          const WAAPIAnimation = generate(WAAPI);
+          const Composed = function (props) {
+            return (
+              <Composable>
+                <WAAPIAnimation hoistAnimation={hoistAnimation}>
+                  {props.children}
+                </WAAPIAnimation>
+              </Composable>
+            );
+          };
+          Composed.propTypes = { children: PropTypes.node };
+          return Composed;
+        },
+        (props) => props.children
+      ),
+    [generators, hoistAnimation]
+  );
+
+  return <ComposedWrapper>{children}</ComposedWrapper>;
+}
+ComposableWrapper.propTypes = {
+  children: PropTypes.node.isRequired,
+  generators: PropTypes.arrayOf(PropTypes.func),
+  hoistAnimation: PropTypes.func.isRequired,
+};
+
+function WAAPIWrapper({ children, target }) {
+  const {
+    state: { getAnimationGenerators },
+    actions: { hoistWAAPIAnimation },
+  } = useStoryAnimationContext();
+
+  return (
+    <ComposableWrapper
+      generators={getAnimationGenerators(target)}
+      hoistAnimation={hoistWAAPIAnimation}
+    >
+      {children}
+    </ComposableWrapper>
+  );
+}
+WAAPIWrapper.propTypes = {
+  children: PropTypes.node,
+  target: PropTypes.string,
+};
```

Comment:
I think `WAAPIWrapper` should allow for a `style` tag to be passed in that gets applied to the outer most container.  If you take a look at how elements are drawn on a completed story in `element.js` (found at: `edit-story/output/element.js`), you'll notice that each element has an explicitly defined `width`, `height`, etc...

I believe we're gonna want our animation wrappers to be the outer most containers on each story element, so we'll be able to pass in the exact styling properties directly in to the wrapper.

What this would mean is in the storybook example you have, instead of needing `SquareWrapper`, it could be written as:
```
<StoryAnimation.WAAPIWrapper
  target="some-id"
  style={{
    position: 'relative',
    marginTop: 20,
    height: 50,
    width: 50,
  }}
>
  <ColorSquare color={state % 2 ? 'tomato' : 'green'} />
</StoryAnimation.WAAPIWrapper>
```

We're guaranteed that style information at the story level so we won't need an extra wrapper around the WAAPIWrapper component if we just apply those styles to the wrapper, what do you think?

Reply:
Also, something else to ponder, we're going to need to be able to set the initial style for an element based off all the animations it has queue up against it.  So if it has a `FadeIn` and `Bounce` animation applied, it'll probably need to have both scale and opacity (properties from different keyframes 😱) set to the proper initial value so when the story plays, things don't just blink out and fade/bounce back in.  They should be hidden, and then appear on hover and stuff (at least on an active story)... but as we are composing the layers, we could be extracting the initial style from each applied animation, and then applying the composite of that style on the outer container 🤔

Reply:
Yea all good points. I think for the first one, I intended `StoryAnimation.WAAPIWrapper` not to be the outer most wrapper, but instead one level down. The current storybook element setup looks like this:
```
<Story>
  ...
  <ElementWrapper>
    <FullSizeWrapper>
      /* Inner Content */
    </FullSizeWrapper>
  </ ElementWrapper>
  ...
</Story>
```
I think it could be nice to minimize roles of everything so the element wrapper is still responsible for positioning while this animation wrappper is only responsible for animating. So I imagined it looking like this:
```
<StoryAnimation.provider animations={...}>
  <Story>
    ...
    <ElementWrapper>
      <StoryAnimation.WAAPIWrapper. target={id}>
        <FullSizeWrapper>
          /* Inner Content */
        </FullSizeWrapper>
      </StoryAnimation.WAAPIWrapper>
    </ ElementWrapper>
    ...
  </Story>
</StoryAnimation.provider>
```

Reply:
> Also, something else to ponder, we're going to need to be able to set the initial style for an element based off all the animations

Are we sure on this? the last animation spec I saw showed the static stories in the final state of the animations, then when you hover it starts over from the start.

I think this is something that if we need to do, we can certainly handle at the animation part level of this setup, but still unsure of the spec here

Reply:
I think you're right for previewing an animation in the dashboard.  But for a generated `amp-story`, I believe the idea is that as you switch from page to page, the page loads in the initial state of the animations, and then plays everything out and stays at the final state... 

`amp-animation` might handle this for us anyway, or at least we can definitely wait until we're integrating directly into stories to see if this is an issue or not 👍

Reply:
Oh and I can see your point of keeping `WAAIWrapper` as having one isolated function, and letting the outer element be the container.  In the same way we can pass in the exact styles to `WAAIWrapper`, we can also pass them into an outer element too, so it ain't but a thang! 👍

Reply:
> But for a generated amp-story, I believe the idea is that as you switch from page to page

yea good point. I guess we'll see if there's any race conditions as we switch to the next page and immediately play it. certainly don't want a frame where the animations in it's final state then flashes to the start of it